std/range/package.d

```diff
@@ -8606,6 +8606,302 @@ if (isForwardRange!Source && hasLength!Source)
+                {
+                    end += sizesCopy.front();
+                    newSizes ~= [sizesCopy.front()];
+                    sizesCopy.popFront();
+                    upper--;
+                }
+
+                import std.algorithm.comparison : min;
+                return varChunks(_source[start .. min(end,_source.length)], newSizes);
+            }
+        }
+
+        // Bidirectional range primitives
+        static if (hasSlicing!Source && hasLength!Source && isForwardRange!Sizes)
+        {
+            @property auto back()
+            {
+                assert(!empty, "back called on empty varChunks");
+
+                size_t start = 0;
+                size_t lastSize = 0;
+                size_t len = _source.length;
+                Sizes sizesCopy = _chunkSizes.save();
+
+                while (!sizesCopy.empty && start < len)
+                {
+                    lastSize = sizesCopy.front();
+                    start += lastSize;
+                    sizesCopy.popFront();
+                }
+
+                assert(start>=len, "Not enough chunkSizes to get the back varChunks");
+
+                return _source[start-lastSize .. len];
+            }
+
+            void popBack()
+            {
+                size_t start = 0;
+                size_t lastSize = 0;
+                size_t len = _source.length;
+                Sizes sizesCopy = _chunkSizes.save();
+
+                while (!sizesCopy.empty && start < len)
+                {
+                    lastSize = sizesCopy.front();
+                    start += lastSize;
+                    sizesCopy.popFront();
+                }
+                assert(start>=len, "Not enough chunkSizes to get the back varChunks length");
+
+                _source = _source[0..(start-lastSize)];
+            }
+        }
+
+    private:
+        Source _source;
+        Sizes _chunkSizes;
+    }
+    else static if (isIntegral!(ElementType!Sizes))
+    {
+
+        import std.typecons : RefCounted;
+
+        static struct Chunk
+        {
+            private RefCounted!Impl impl;
+
+            @property bool empty() { return impl.curSizeLeft == 0 || impl.r.empty || impl.s.empty; }
+            @property auto front() { return impl.r.front; }
+            void popFront()
+            {
+                assert(impl.curSizeLeft > 0 && !impl.r.empty);
+                impl.curSizeLeft--;
+                impl.r.popFront();
+            }
+        }
+
+        static struct Impl
+        {
+            private Source r;
+            private Sizes s;
+            private size_t curSizeLeft;
+        }
+
+        private RefCounted!Impl impl;
+
+        private this(Source r, Sizes s)
+        {
+            impl = RefCounted!Impl(r, s, r.empty ? 0 : (s.empty ? 0 : s.front));
+        }
+
+        @property bool empty() { return impl.r.empty || impl.s.empty; }
+        @property Chunk front() return { return Chunk(impl); }
+
+        void popFront()
```

Comment:
This `popFront` implementation invalidates the last returned `front` while the previous version probably keeps the last element valid. That is quite dangerous and should at least be documented.